table/tables/index.go

```diff
@@ -222,25 +304,45 @@ func (c *index) Create(sctx sessionctx.Context, rm kv.RetrieverMutator, indexedV
 
 	// save the key buffer to reuse.
 	writeBufs.IndexKeyBuf = key
-	if !distinct {
-		// non-unique index doesn't need store value, write a '0' to reduce space
-		value := []byte{'0'}
+	var idxVal []byte
+	if collate.NewCollationEnabled() && !c.checkContainNonBinaryString() {
+		idxVal = make([]byte, 1+len(restoredValue))
```

Comment:
Considering the performance, all bytes allocation can be done at once to avoid the extra alloc-and-copy by `append()`. 

However, the optimization can be done in another PR